### PR TITLE
[fix][offload] Fix Offload readHandle cannot close multi times.

### PR DESCRIPTION
**Pull request by dao-jun**
Fixes https://github.com/apache/pulsar/pull/22035

Main Issue: #xyz

PIP: #xyz 

### Motivation

Make `FileStoreBackedReadHandleImpl`, `BlobStoreBackedReadHandleImpl` and `BlobStoreBackedReadHandleImplV2`
`closeAsync` method can be called multi times. 

### Verifying this change

- [ ] Make sure that the change passes the CI checks.

*(Please pick either of the following options)*

This change is a trivial rework / code cleanup without any test coverage.

*(or)*

This change is already covered by existing tests, such as *(please describe tests)*.

*(or)*

This change added tests and can be verified as follows:

*(example:)*
  - *Added integration tests for end-to-end deployment with large payloads (10MB)*
  - *Extended integration test for recovery after broker failure*

### Does this pull request potentially affect one of the following parts:

*If the box was checked, please highlight the changes*

- [ ] Dependencies (add or upgrade a dependency)
- [ ] The public API
- [ ] The schema
- [ ] The default values of configurations
- [ ] The threading model
- [ ] The binary protocol
- [ ] The REST endpoints
- [ ] The admin CLI options
- [ ] The metrics
- [ ] Anything that affects deployment

### Documentation

- [ ] `doc` 
- [ ] `doc-required` 
- [x] `doc-not-needed` 
- [ ] `doc-complete` 

### Matching PR in forked repository

PR in forked repository: https://github.com/dao-jun/pulsar/pull/9

### CODE REVIEW

File: pulsar-common/src/main/java/org/apache/pulsar/common/util/FutureUtil.java [hunk 1]
```diff
@@ -347,4 +347,21 @@ public static void safeRunAsync(Runnable runnable,
                     return null;
                 });
     }
+
+    public static <T> @Nonnull CompletableFuture<T> apply(@Nonnull CompletableFuture<T> s,
```
Comment (lhotari) [inline pulsar-common/src/main/java/org/apache/pulsar/common/util/FutureUtil.java:351]: please add a proper name and javadoc for this function. `completeAfter` would be a better name than `apply`.
a good guideline in naming is to be rather specific than generic. `apply` is very vague.

File: tiered-storage/file-system/src/main/java/org/apache/bookkeeper/mledger/offload/filesystem/impl/FileStoreBackedReadHandleImpl.java [hunk 2]
```diff
@@ -93,14 +103,20 @@ public LedgerMetadata getLedgerMetadata() {
     @Override
     public CompletableFuture<Void> closeAsync() {
         CompletableFuture<Void> promise = new CompletableFuture<>();
+
+        if (!closeFuture.compareAndSet(null, promise)) {
+            return FutureUtil.apply(closeFuture.get(), promise);
+        }
```
Comment (lhotari) [inline tiered-storage/file-system/src/main/java/org/apache/bookkeeper/mledger/offload/filesystem/impl/FileStoreBackedReadHandleImpl.java:109]: why isn't `closeFuture.get()` returned directly?

**Review by lhotari — CHANGES_REQUESTED** [review]

File: tiered-storage/file-system/src/main/java/org/apache/bookkeeper/mledger/offload/filesystem/impl/FileStoreBackedReadHandleImpl.java (hunk 2 above)
Comment (dao-jun) [reply tiered-storage/file-system/src/main/java/org/apache/bookkeeper/mledger/offload/filesystem/impl/FileStoreBackedReadHandleImpl.java:109]: I also thought about returning `closeFuture.get()` directly, but `promise` is already created, I want to use it

File: pulsar-common/src/main/java/org/apache/pulsar/common/util/FutureUtil.java (hunk 1 above)
Comment (dao-jun) [reply pulsar-common/src/main/java/org/apache/pulsar/common/util/FutureUtil.java:351]: you are right(naming is difficult...)

File: tiered-storage/file-system/src/main/java/org/apache/bookkeeper/mledger/offload/filesystem/impl/FileStoreBackedReadHandleImpl.java (hunk 2 above)
Comment (lhotari) [reply tiered-storage/file-system/src/main/java/org/apache/bookkeeper/mledger/offload/filesystem/impl/FileStoreBackedReadHandleImpl.java]: > but promise is already created, I want to use it

Why? The creation is an extremely lightweight operation. That's fine to let GC take care of that. Very short living objects like this are handled extremely efficiently by Java GC. There's no need to worry about that.

Comment (dao-jun) [reply tiered-storage/file-system/src/main/java/org/apache/bookkeeper/mledger/offload/filesystem/impl/FileStoreBackedReadHandleImpl.java]: Yes, I known, just don't want to waste it. 
return `closeFuture.get()` is better, that I don't need to add a new method to `FutureUtil`

File: pulsar-common/src/main/java/org/apache/pulsar/common/util/FutureUtil.java [hunk 3]
```diff
@@ -348,20 +348,28 @@ public static void safeRunAsync(Runnable runnable,
                 });
     }
 
-    public static <T> @Nonnull CompletableFuture<T> apply(@Nonnull CompletableFuture<T> s,
-                                                          @Nonnull CompletableFuture<T> d) {
-        if (s == d) {
-            return d;
+    /**
+     * Apply source's result to target.
```
[the remark this answers is not in the document]
Comment (dao-jun) [reply pulsar-common/src/main/java/org/apache/pulsar/common/util/FutureUtil.java]: removed the method

**Review by lhotari — APPROVED** [review]
LGTM. Good work @dao-jun

File: tiered-storage/jcloud/src/main/java/org/apache/bookkeeper/mledger/offload/jcloud/impl/BlobStoreBackedReadHandleImplV2.java [hunk 4]
```diff
@@ -123,8 +125,13 @@ public LedgerMetadata getLedgerMetadata() {
...
     @Override
     public CompletableFuture<Void> closeAsync() {
-        CompletableFuture<Void> promise = new CompletableFuture<>();
+        if (closeFuture.get() != null || !closeFuture.compareAndSet(null, new CompletableFuture<>())) {
+            return closeFuture.get();
+        }
+
+        CompletableFuture<Void> promise = closeFuture.get();
         executor.execute(() -> {
+
```
Comment (Technoboy-) [inline tiered-storage/jcloud/src/main/java/org/apache/bookkeeper/mledger/offload/jcloud/impl/BlobStoreBackedReadHandleImplV2.java:134]: Remove this blank and help add at line 153
```
if (log.isDebugEnabled()) 
```

Comment (dao-jun) [reply tiered-storage/jcloud/src/main/java/org/apache/bookkeeper/mledger/offload/jcloud/impl/BlobStoreBackedReadHandleImplV2.java]: @Technoboy- fixed

**Comment by dao-jun** [comment]
all the ci checks passed, merging...

**Pull request merged**